blog/_posts/2022-12-23-wildcart-domain-with-caddy-for-tunnel-hosting.md

```diff
@@ -0,0 +1,137 @@
+sudo systemctl enable --now inlets
+```
+
+Verify that the service is running:
+
+```bash
+systemctl status inlets
+```
+
+## Setup Caddy
+
+Caddy will be used as a reverse proxy for the inlets http tunnel server. It will   
+
+By default the Caddy server does not contain any DNS modules. These can be added to your download from [caddyserver.com](https://caddyserver.com/download). Alternatively you can use xcaddy to build caddy with your DNS provider plugged in. The [caddy community wiki](https://caddy.community/t/how-to-use-dns-provider-modules-in-caddy-2/8148) has more details on how to enable the DNS challenge for your provider.
+
+Create a DNS A record for the wildcard domain you want to use and have it point to the public IP address of your server.
+
+Next create a `Caddyfile`:
+
+```
+{
+    email email@example.com
+}
+
+exit.example.com {
+    reverse_proxy localhost:8123
+}
+
+*.exit.example.com {
+    tls {
+        dns cloudflare {env.CF_API_TOKEN}
+    }
+    reverse_proxy localhost:8000
+}
+```
+
+We define three blocks in the caddy configuration file. The first block on the top of the file can be used to define the email to use for the renewal notifications on the domains.
+
+In the second we configure caddy to reverse proxy the configuration plane for the tunnel server using the domain `exit.example.com`. Caddy will automatically obtain a valid TLS certificate for this domain.
+
+In the third block we configure a wildcard domain `*.exit.example.com` and point it to the data plane of the tunnel. To generate a wildcard certificate you will need to use the DNS-01 challenge type which requires using a supported DNS provider. This is defined with a `tls { }` block added below the domain definition. The configuration might be different depending on you DNS provider. You can find the specific configuration info and instructions to configure authentication for you provider in the [repo of you DNS provider plugin](https://github.com/orgs/caddy-dns/repositories?type=all).
+
+In this example we use the [cloudflare DNS provider](https://github.com/caddy-dns/cloudflare).
+
+You can replace {env.CF_API_TOKEN} with the actual auth token if you prefer to put it directly in your config instead of an environment variable.
+
+Make sure to move the caddy binary you downloaded or compiled using xcaddy into you path:
+
+```bash
+sudo mv caddy /usr/bin/
+```
+
+Run the caddy server:
+
+```bash
+caddy run --config ./Caddyfile
+```
+
+You can [follow the instructions in the documentation](https://caddyserver.com/docs/running) to run the caddy server as a systemd service.
+
+## Connect the tunnel client
+
+You can use the client to connect to you server and specify a subdomain for each local service such as:
+
+```bash
+inlets-pro http client \
+  --token $TOKEN \
+  --url=wss://exit.example.com \
+  --upstream=blog1.exit.example.com=http://127.0.0.1:8080 \
+  --upstream=api.exit.example.com=http://127.0.0.1:6000 \
+  --auto-tls=false
+```
+
```

Comment:
Let's have a systemd unit file for the server with --generate=system

Plus a systemd unit file for Caddy, so people can reboot the machine without logging in.

Also show that two clients can connect from different machines or at different times one exposing blog and the other exposing... "api"?